File: zig/examples/words_geometry.zig
// words_geometry — build a PDF from Markdown, open it, extract word geometry.
//
// Shared regression scenario (mirrored across language bindings). Exits
// non-zero on any failed assertion; prints "WORDS OK" on success.
const std = @import("std");
const pdf_oxide = @import("pdf_oxide");

pub fn main() !void {
    const a = std.heap.page_allocator;

    var pdf = try pdf_oxide.Pdf.fromMarkdown(
        "# Hello pdf_oxide\n\nThis is a **Zig** regression example.\n",
    );
    defer pdf.deinit();

    const bytes = try pdf.toBytes(a);
    defer a.free(bytes);

    var doc = try pdf_oxide.Document.openFromBytes(bytes);
    defer doc.deinit();

    const words = try doc.extractWords(a, 0);
    defer pdf_oxide.Document.freeWords(a, words);

    var buf: [4096]u8 = undefined;
    var fw = std.fs.File.stdout().writer(&buf);
    const stdout = &fw.interface;

    try stdout.print("word count: {d}\n", .{words.len});

    if (words.len == 0) {
        try stdout.print("assertion failed: no words extracted\n", .{});
        try stdout.flush();
        std.process.exit(1);
    }

    const first = words[0];
    try stdout.print("first word: \"{s}\"  bbox=({d}, {d}, {d}, {d})\n", .{
        first.text, first.bbox.x, first.bbox.y, first.bbox.width, first.bbox.height,
    });

    if (!std.mem.eql(u8, first.text, "Hello")) {
        try stdout.print("assertion failed: first word is not \"Hello\"\n", .{});
        try stdout.flush();
        std.process.exit(1);
    }
    if (!(first.bbox.width > 0.0 and first.bbox.height > 0.0)) {
        try stdout.print("assertion failed: first word has no bbox\n", .{});
        try stdout.flush();
        std.process.exit(1);
    }

    try stdout.print("WORDS OK\n", .{});
    try stdout.flush();
}

File: objc/examples/words_geometry.m
// words_geometry — build a PDF from Markdown, then extract positioned words.
// Shared-scenario regression example. Returns non-zero on assertion failure.
#import "POXPdfOxide.h"
#import <Foundation/Foundation.h>

int main(void) {
    @autoreleasepool {
        NSError* err = nil;
        POXPdf* pdf = [POXPdf fromMarkdown:@"# Hello pdf_oxide\n\nThis is a "
                                           @"**Objective-C** regression example.\n"
                                     error:&err];
        if (!pdf) {
            fprintf(stderr, "error: %s\n", err.localizedDescription.UTF8String);
            return 1;
        }

        POXDocument* doc = [POXDocument openFromBytes:[pdf toBytesWithError:&err]
                                                error:&err];
        if (!doc) {
            fprintf(stderr, "error: %s\n", err.localizedDescription.UTF8String);
            return 1;
        }

        NSArray<POXWord*>* words = [doc extractWords:0 error:&err];
        if (!words) {
            fprintf(stderr, "error: %s\n", err.localizedDescription.UTF8String);
            return 1;
        }
        if (words.count == 0) {
            fprintf(stderr, "assertion failed: no words extracted\n");
            return 1;
        }

        POXWord* first = words[0];
        printf("words: %lu\n", (unsigned long)words.count);
        printf("first word: \"%s\"\n", first.text.UTF8String);
        printf("bbox: x=%f y=%f w=%f h=%f\n", first.bbox.x, first.bbox.y,
               first.bbox.width, first.bbox.height);

        if (![first.text isEqualToString:@"Hello"]) {
            fprintf(stderr, "assertion failed: first word != \"Hello\"\n");
            return 1;
        }
        if (!(first.bbox.width > 0 && first.bbox.height > 0)) {
            fprintf(stderr, "assertion failed: first word has no bbox\n");
            return 1;
        }
        printf("WORDS OK\n");
        return 0;
    }
}
